### PR TITLE
Add Upstash Redis caching with local Redis fallback and per-user cache invalidation

### PR DESCRIPTION
**Pull request by JoshCLWren**
## Summary

Add Redis caching (via `@cached` decorator) to all high-frequency read queries with precise, user-scoped cache invalidation on mutations.

### Caching Infrastructure
- `app/cache.py`: `UpstashCache` now supports both Upstash cloud (`UPSTASH_REDIS_REST_URL`+`UPSTASH_REDIS_REST_TOKEN`) and local Redis (`REDIS_URL`) via `redis-py`
- `app/config.py`: Added `redis_url` field to `RedisSettings`
- `app/main.py`: Startup event initializes cache with Upstash or local Redis

### Decorated Endpoints
- Thread: `list_threads`, `get_thread`
- Collection: `list_collections`, `get_collection`
- Issue: `list_issues`, `get_issue`, `validate_issue_order`
- Session: `get_current_session`, `list_sessions`, `get_session`, `get_session_details`, `get_session_snapshots`
- Dependency: `list_thread_dependencies`, `list_issue_dependencies`, `get_thread_blocking_info`, `get_dependency`, `check_thread_dependency_order`, `get_thread_connected_threads`

### Invalidation
- Every mutation now invalidates only the **current user's caches** (e.g., `f\"cache:list_threads:User:{user.id}:*\"`)
- All concurrent invalidations use `asyncio.gather`
- Removed all `clear_cache=None` globals and the `get_current_session_cached` hack

### Bug Fix
- `refresh_user_blocked_status` now invalidates its own cache before reading blocked thread IDs (fixes stale read issue)

### Dev/Test Infrastructure
- Added `redis>=5.0.0` dependency
- Redis container in `docker-compose.test.yml`
- `REDIS_URL` in `.env.test`
- Test fixtures that initialize and flush cache between tests
- Redis service + env in all CI test jobs

## Testing
- ✅ All 726 backend tests pass
- ✅ 6 E2E API tests pass  
- ✅ Ruff lint clean
- ✅ `ty` typecheck clean

## Summary by CodeRabbit

* **New Features**
  * Added Redis/Upstash-backed caching with TTL tiers across threads, collections, dependencies, issues, sessions, and blocking info.
  * Cache is now automatically invalidated after key mutations (edits, ratings, queue changes, snooze/unsnooze, undo, etc.).
  * Added Redis configuration and initialized/shutdown cache on app startup.
* **Performance**
  * Faster repeat reads via cached GET responses with consistent expiration behavior and safer handling when Redis is unavailable.
* **Testing**
  * Added Redis-enabled test fixtures and a full cache regression suite.
* **Chores**
  * Updated CI and Docker-based tests to provision and expose Redis for backend and E2E runs.

### CODE REVIEW

File: app/api/thread.py [hunk 1]
```diff
@@ -39,6 +40,14 @@
 logger = logging.getLogger(__name__)
 
 
+async def _invalidate_thread_caches(user_id: int, thread_id: int | None = None) -> None:
+    """Invalidate thread cache entries for a specific user."""
+    coros = [invalidate_cache(f"cache:list_threads:User:{user_id}:*")]
+    if thread_id is not None:
+        coros.append(invalidate_cache(f"cache:get_thread:{thread_id}:User:{user_id}:"))
+    await asyncio.gather(*coros)
```
Comment (JoshCLWren) [inline app/api/thread.py:56]: **Blocking: session mutations in this module still bypass session invalidation.** `set_pending_thread()` writes a roll event and changes `Session.pending_thread_id`, but returns after `db.commit()` without clearing `get_current_session` or the other session caches. The current Firefox failure where `beforeSession.active_thread` is undefined maps directly to that path. `delete_thread()` also clears a pending thread in the database without clearing session caches.

Please call `_invalidate_session_caches(current_user.id)` after those commits and add a warm-cache regression test: cache `/api/sessions/current/`, call `set-pending`, repeat the GET, and assert the selected thread is active.

File: app/cache.py [hunk 2]
```diff
@@ -0,0 +1,512 @@
...
+        - Local Redis: provide local_url (e.g., redis://localhost:6379/0)
+        """
+        if self._initialized:
+            logger.warning("Cache already initialized")
+            return
+
+        if local_url:
+            try:
+                self._client = aioredis.Redis.from_url(local_url, decode_responses=True)
+                await self._client.ping()
```
Comment (JoshCLWren) [inline app/cache.py]: **Blocking: successful reconnection should reset the circuit breaker.** The test fixture closes and reinitializes this singleton, but the breaker state survives. If an earlier cross-loop/cache operation opens it, this `PING` succeeds while subsequent `get()` and `set()` still short-circuit to `None`/`False`. That is consistent with the two new backend test failures despite Redis logging a successful initialization.

Add an explicit `CircuitBreaker.reset()` that restores `CLOSED`, zero failures, and clears `_opened_at`, then call it after each successful initialization ping. The direct cache tests should also assert that `await cache.set(...) is True` so failures surface at the write.

File: app/api/queue.py [hunk 3]
```diff
@@ -102,8 +102,10 @@ async def move_thread_position(
         await db.commit()
         await db.refresh(thread)
 
-    if clear_cache:
-        clear_cache()
+    await asyncio.gather(
```
Comment (JoshCLWren) [inline app/api/queue.py]: **Major: queue changes also invalidate cached session representations.** `SessionResponse.active_thread.queue_position` is built from the current `Thread`, so move-to-position, move-front, and move-back can leave `get_current_session`, `get_session`, and session lists stale. Please include `_invalidate_session_caches(current_user.id)` here and in the sibling move handlers.

The same applies to shuffle. Shuffle should also clear `cache:get_thread:*:User:{user_id}:` because it changes every active thread's position, not only the list ordering.

File: app/cache.py [hunk 4]
```diff
@@ -0,0 +1,512 @@
...
+        if self._client is None:
+            return False
+
+        try:
+            prepared = UpstashCache._prepare_value(value)
+            data = json.dumps(
+                prepared,
+                default=str,
+            )
+            if ttl:
```
Comment (JoshCLWren) [inline app/cache.py]: **Minor correctness issue: `ttl=0` still means no expiration.** This truthiness check routes zero to the no-`ex` branch, creating a persistent key. Use `if ttl is not None:` so an explicit zero is passed through rather than silently changing its meaning. Please add a focused test for the configured zero-TTL case.

**Review by JoshCLWren — COMMENTED** [review]
**Merge-blocking review:** GitHub does not allow a pull-request author to submit `Request changes` on their own PR, so this is posted as a Comment review instead. The repair commit fixes the central cache-key mismatch and several important stale-read paths. The remaining failures now point to a smaller set of concrete lifecycle and invalidation gaps. Please address the inline findings before merge, then rerun the backend and Playwright suites.
